### ISSUE TITLE
ActiveSupport::Duration#+, -, since and ago should return Time consistently for Date object

### ISSUE THREAD

**Issue by JunichiIto** [rails/rails#37450]
### Steps to reproduce

```ruby
date = Date.today
date + 1.hour
date + 0.hours
1.hour.since(date)
0.hours.since(date)
```

### Expected behavior

```ruby
date = Date.today
#=> 2019-10-11

# returns Time
date + 1.hour
#=> 2019-10-11 01:00:00 +0000

# returns Time
date + 0.hours
#=> 2019-10-11 00:00:00 +0000

# returns Time
1.hour.since(date)
#=> 2019-10-11 01:00:00 +0000

# returns Time
0.hours.since(date)
#=> 2019-10-11 00:00:00 +0000
```

### Actual behavior

```ruby
date = Date.today
#=> 2019-10-11

# returns Time
date + 1.hour
#=> 2019-10-11 01:00:00 +0000

# returns Date (wrong)
date + 0.hours
#=> 2019-10-11

# returns Time
1.hour.since(date)
#=> 2019-10-11 01:00:00 +0000

# returns Date (wrong)
0.hours.since(date)
#=> 2019-10-11
```

`-` and `ago` also return Date with zero duration.

### Note

With Rails 5.1, `date + 0.hours` and `0.hours.since(date)` return Time.

After investigation, I found this behavior was introduced since Rails 5.2 by the PR below:

https://github.com/rails/rails/pull/31310

When we upgraded our app from 5.1 to 5.2, we noticed the following result changed:

```ruby
Time.zone.name
#=> Tokyo (JST +09:00, not UTC)

date = '2019-10-11'.to_date
offset = 0 # user input value 
deadline = '2019-10-11'.in_time_zone

# Compare values
(date + offset.hours) <= deadline
#=> true (with Rails 5.1)
#=> false (with Rails 5.2)
```

So I think it should be reverted.

### System configuration
**Rails version**: 5.2+

**Ruby version**: 2.6.5

**Comment by JunichiIto** [comment]
Anyone?

**Comment by kamipo** [comment]
Fixed by #37882.
